rb_link/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+#[derive(Serialize, Deserialize)]
+pub struct B2RMessage {
+    pub id: u32,
+    pub cycles: u32,
+    pub message: Vec<u8>,
+}
+
+/// A server for interacting with Bluesim.
+/// Cache bidirectional data and send data upon receiving requests.
+pub struct B2RServer {
+    // probe_infos: Arc<Mutex<HashMap<u32, ProbeInfo>>>,
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+    r2b_cache: Arc<Mutex<HashMap<u32, VecDeque<R2BMessage>>>>,
+}
+
+impl Default for B2RServer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl B2RServer {
+    /// make a new server
+    pub fn new() -> B2RServer {
+        B2RServer {
+            // probe_infos: Arc::new(Mutex::new(HashMap::new())),
+            b2r_cache: Arc::new(Mutex::new(HashMap::new())),
+            r2b_cache: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+    // pub fn add_probe(&mut self, id: u32, get_t_width: u32, put_t_width: u32) {
+    //     let probe_info = ProbeInfo {
+    //         id,
+    //         get_t_width,
+    //         put_t_width,
+    //     };
+    //     self.probe_infos.lock().unwrap().insert(id, probe_info);
+    // }
+
+    /// Start a thread to run the server.
+    /// Create a UnixListener at "/tmp/b2rr2b".
+    /// Return the JoinHandle of that thread.
+    /// This function needs to be called before running your Bluesim program.
+    pub fn serve(&mut self) -> JoinHandle<()> {
+        // let probe_infos = self.probe_infos.clone();
+        let b2r_cache = self.b2r_cache.clone();
+        let r2b_cache = self.r2b_cache.clone();
+        thread::spawn(move || {
+            let _ = fs::remove_file("/tmp/b2rr2b");
+            let listener = UnixListener::bind("/tmp/b2rr2b").expect("Failed to bind Unix listener");
+            for stream in listener.incoming() {
+                let mut stream = match stream {
+                    Ok(stream) => stream,
+                    Err(_) => todo!(),
+                };
+                let message = match receive_getput(&mut stream) {
+                    Ok(m) => m,
+                    Err(_) => todo!(),
+                };
+                match message {
+                    GetPutMessage::Get(id) => {
+                        // println!("receive get from id {}", id);
+                        loop {
+                            let mut r2b_cache = r2b_cache.lock().unwrap();
+                            if let Some(queue) = r2b_cache.get_mut(&id) {
+                                if let Some(r2b_message) = queue.pop_front() {
+                                    stream.write_all(&r2b_message.message).unwrap();
+                                    drop(r2b_cache);
+                                    break;
+                                }
+                            }
+                            drop(r2b_cache);
+                            thread::sleep(Duration::from_millis(100));
```

Comment:
it's not good to use sleep, simulator runs faster than you think, so you sleep here may slowdown the simulator. Try use events or signals to wait.